firedrake/function.py

```diff
@@ -304,14 +304,14 @@ def __dir__(self):
         return list(OrderedDict.fromkeys(dir(self._data) + current))
 
     @utils.cached_property
+    @FunctionMixin._ad_annotate_subfunctions
     def subfunctions(self):
         r"""Extract any sub :class:`Function`\s defined on the component spaces
         of this this :class:`Function`'s :class:`.FunctionSpace`."""
         return tuple(type(self)(V, val)
                      for (V, val) in zip(self.function_space(), self.topological.subfunctions))
 
-    @PETSc.Log.EventDecorator()
-    @FunctionMixin._ad_annotate_split
+    @FunctionMixin._ad_annotate_subfunctions
```

Comment:
When we annotate a function it is modified to accept a few new kwargs. pyadjoint tests were failing because not decorating the deprecated function meant that these additional kwargs weren't recognised.

I should probably revisit the `zero` one to pass `args` and `kwargs` through to `assign`.

Reply:
I.e. it didn't know about `.subfunctions` and also doesn't know about `.zero`?

Reply:
pyadjoint is still using the deprecated `split` so yeah it doesn't know about `subfunctions`. Adding a pyadjoint test for `zero` is now on my TODO list.